meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java

```diff
@@ -75,9 +76,12 @@ public DistributedSemaphore(MetaClientInterface<DataRecord> client) {
     _metaClient = client;
     try {
       _metaClient.connect();
-      // TODO: Differentiate exception catch between already connected and already closed.
     } catch (IllegalStateException e) {
-      // Ignore as it either has already been connected or already been closed.
+      if (e.getMessage().contains("connect() has already been called")) {
+        LOG.info("Client already connected");
```

Comment:
Checking test message is a hacky way. Any other options?

Reply:
Right but there is no other way as far as I know. There is a method that checks whether a node is closed but that is zk specific and shouldn't be added in a generic distributedSemaphore client creation. Zk.connect() throws the same IllegalStateException for both client exceptions so there isn't a way of separating the two. Furthermore, while kinda of hacky, the current implementation of zkClient.connect() exceptions is very mature and the chances of one modifying it is slim. However, this begs the question for future use cases besides zk.

Reply:
How about we add an API `isClosed` in metaclientInterface. So that we can use this generic API in recipes client.

Reply:
+1 to xyuanlu's point
Alternatively, you can make `_metaClient.connect()` idempotent, all following calls after initial connect() are no-op. Users won't have to worry about checking isConnected. This might lead to some behavior changes though.